Algorithm.CSharp/CustomDataUnderlyingOptionSymbolMappingRegressionAlgorithm.cs

```diff
@@ -0,0 +1,147 @@
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect;
+using QuantConnect.Data;
+using QuantConnect.Data.Custom.SEC;
+using QuantConnect.Interfaces;
+
+namespace QuantConnect.Algorithm.CSharp
+{
+    /// <summary>
+    /// Regression algorithm ensures that mapping is also applied to the underlying symbol(s) for custom data subscriptions
+    /// </summary>
+    /// <meta name="tag" content="using data" />
+    /// <meta name="tag" content="custom data" />
+    /// <meta name="tag" content="regression test" />
+    /// <meta name="tag" content="rename event" />
+    /// <meta name="tag" content="map" />
+    /// <meta name="tag" content="mapping" />
+    /// <meta name="tag" content="map files" />
+    public class CustomDataUnderlyingOptionSymbolMappingRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
+    {
+        private bool _initialSymbolChangedEvent;
+
+        // Option to add custom data with as Symbol
+        private Symbol _optionSymbol;
+
+        // Custom data that was added with option ticker
+        private Symbol _customDataOptionSymbol;
+
+
+        /// <summary>
+        /// Adds option NWSA -&gt; FOXA so that we can test if mapping occurs to the underlying symbols in the custom data subscription
+        /// </summary>
+        public override void Initialize()
+        {
+            SetStartDate(2013, 6, 28);
+            SetEndDate(2013, 7, 02);
+            SetCash(100000);
+
+            _optionSymbol = AddOption("FOXA", Resolution.Daily).Symbol;
+            _customDataOptionSymbol = AddData<SECReport10K>(_optionSymbol).Symbol;
+        }
+
+        /// <summary>
+        /// Checks that custom data underlying symbols match the expected symbols and contains chain of custom -&gt; option -&gt equity
+        /// </summary>
+        /// <param name="data"></param>
+        public override void OnData(Slice data)
+        {
+            if (data.SymbolChangedEvents.Any() && !_initialSymbolChangedEvent)
+            {
+                _initialSymbolChangedEvent = true;
+                return;
+            }
+
+            if (data.SymbolChangedEvents.Any())
+            {
+                if (data.SymbolChangedEvents.ContainsKey(_customDataOptionSymbol) && data.SymbolChangedEvents.ContainsKey(_optionSymbol))
+                {
+                    var expectedUnderlying = "?FOXA";
+                    var underlying = data.SymbolChangedEvents.Keys.Where(x => x.SecurityType == SecurityType.Base && x == _customDataOptionSymbol).Single().Underlying;
+                    var symbol = data.SymbolChangedEvents.Keys.Where(x => x.SecurityType == SecurityType.Equity && x == _optionSymbol).Single();
+
+                    if (SubscriptionManager.Subscriptions.Where(x => (x.SecurityType == SecurityType.Base || x.SecurityType == SecurityType.Option || x.SecurityType == SecurityType.Equity) && x.MappedSymbol == expectedUnderlying).Count() != 3)
```

Comment:
nit - Both `Count` and `Single` can accept a `predicate` so the `Where` statement can be removed - Idem for the other csharp algo